doc/cprover-manual/modeling-shadow-memory.md

```diff
@@ -0,0 +1,148 @@
+# Shadow Memory
+
+(The original artefact's implementation is outlined in the paper [CBMC-SSM:
+Bounded Model Checking of C Programs with Symbolic Shadow
+Memory](https://dl.acm.org/doi/abs/10.1145/3551349.3559523).)
+
+## Introduction
+
+CBMC contains a component that supports *Symbolic Shadow Memory*. Shadow memory
+allows one to create a parallel memory structure that is maintained by the
+analysis program and that shadows the standard memory of the program. In that
+memory structure, a user can create fields, for which he can later set and
+retrieve values.
+
+By doing so, one can organise his own tracking of metadata related to the code,
+in a way that is then considered by the backend of CBMC during analysis. This
+can be used to implement novel analyses that the CBMC framework itself doesn't
+support - for example, the paper above presents as an example a taint analysis
+implemented with the use of the SSM component described here.
+
+## Usage
+
+A user can interact with the Symbolic Shadow Memory component (from now on
+onwards, SSM) through four CBMC primitives, which allow one to declare shadow
+memory fields, and retrieve/set their corresponding values:
+
+* `__CPROVER_field_decl_local`
+* `__CPROVER_field_decl_global`
+* `__CPROVER_get_field`
+* `__CPROVER_set_field`
+
+More precisely, their signatures (in pseudo-C, because of some constraints that
+we can't express using the type system), along with some small examples of their
+usage, look like this:
+
+### `void __CPROVER_field_decl_local(type1 field_name, SSM_value_type init_value)`
+
+Type constraints:
+
+* `type1`: string literal, such as `"field"`,
```

Comment:
Maybe worth emphasising that by string literal we mean that the field name must be compile time constant and may not be the result of runtime string manipulation. That is, if my understanding of what you are trying to say here is correct...

Reply:
What I meant is that it has to be a string literal such as `"this is a string literal"`.

In the `exprt` world it has to be a `constant_stringt`.

Reply:
Ok. My understanding is correct. I think it is worth making the documentation spell out the exact details of what is and is not allowed as an argument here.